File: src/jsonrpc/eth/createAccessList/eth_createAccessList.zig
const std = @import("std");
const types = @import("../../types.zig");

/// Generates an access list for a transaction.
///
/// Example:
/// Transaction: ...
/// Block: "latest"
/// Result: ...
///
/// Implements the `eth_createAccessList` JSON-RPC method.
pub const EthCreateAccessList = @This();

/// The JSON-RPC method name
pub const method = "eth_createAccessList";

/// Parameters for `eth_createAccessList`
pub const Params = struct {
    /// Transaction object generic to all types
    transaction: types.Quantity,
    /// Block number or tag
    block: types.Quantity,

    pub fn jsonStringify(self: Params, jws: *std.json.Stringify) !void {
        try jws.beginArray();
        try jws.write(self.transaction);
        try jws.write(self.block);
        try jws.endArray();
    }

    pub fn jsonParseFromValue(allocator: std.mem.Allocator, source: std.json.Value, options: std.json.ParseOptions) !Params {
        if (source != .array) return error.UnexpectedToken;
        if (source.array.items.len != 2) return error.InvalidParamCount;

        return Params{
            .transaction = try std.json.innerParseFromValue(types.Quantity, allocator, source.array.items[0], options),
            .block = try std.json.innerParseFromValue(types.Quantity, allocator, source.array.items[1], options),
        };
    }
};

/// Result for `eth_createAccessList`
pub const Result = struct {
    /// Access list result
    value: types.Quantity,

    pub fn jsonStringify(self: Result, jws: *std.json.Stringify) !void {
        try jws.write(self.value);
    }

    pub fn jsonParseFromValue(allocator: std.mem.Allocator, source: std.json.Value, options: std.json.ParseOptions) !Result {
        return Result{
            .value = try std.json.innerParseFromValue(types.Quantity, allocator, source, options),
        };
    }
};

File: src/jsonrpc/eth/createAccessList/eth_createAccessList.js
/**
 * @fileoverview eth_createAccessList JSON-RPC method
 */

/**
 * @typedef {import('../../../primitives/Address/AddressType.js').AddressType} Address
 * @typedef {import('../../types/index.js').Hash} Hash
 * @typedef {import('../../types/index.js').Quantity} Quantity
 * @typedef {import('../../types/index.js').BlockTag} BlockTag
 * @typedef {import('../../types/index.js').BlockSpec} BlockSpec
 * @typedef {import('../../../provider/types.js').RequestArguments} RequestArguments
 * @typedef {import('../call/eth_call.js').CallParams} CallParams
 */

/**
 * Generates an access list for a transaction.
 *
 * @example
 * Transaction: ...
 * Block: "latest"
 * Result: ...
 *
 * Implements the `eth_createAccessList` JSON-RPC method.
 */

/** The JSON-RPC method name */
export const method = "eth_createAccessList";
/**
 * Result for `eth_createAccessList`
 *
 * Access list result
 *
 * @typedef {Quantity} Result
 */

/**
 * Creates a eth_createAccessList JSON-RPC request
 *
 * @param {CallParams} address
 * @param {BlockSpec} [block]
 * @returns {RequestArguments}
 */
export function CreateAccessListRequest(address, block = "latest") {
	return { method, params: [address, block] };
}
